### PR TITLE
Significant refactoring and simplification.

### PR DESCRIPTION
**Pull request by SeanRBurton**
Does what it says on the tin.

### CODE REVIEW

File: src/arbitrary.rs [hunk 1]
```diff
@@ -20,6 +20,7 @@ use rand::Rng;
 /// `gen` function in this crate.
 pub trait Gen : Rng {
     fn size(&self) -> usize;
+    fn reset(&mut self) {}
```
Comment (BurntSushi) [inline src/arbitrary.rs]: Can you please explain this? Preferably with documentation. (Bonus points for documenting `size`.)

**Comment by BurntSushi** [comment]
@SeanRBurton This is a heroic clean up. I tried myself a few different times to get the shrinker impl'd in a macro so that we could make this work easily with multiple arities. Really nice job.

I think I have one doc nit and then a rebase and this should be good to go. Exciting!

**Comment by SeanRBurton** [comment]
I merged this with master. It should now be good to go.

**Comment by BurntSushi** [comment]
Yay! Awesome work. Thanks so much.

**Pull request merged**